services/streammanager/googlecloudfunction/googlecloudfunction.go

```diff
@@ -90,153 +85,80 @@ func (c *GoogleCloudFunctionClientImpl) GetToken(ctx context.Context, functionUr
-
-	respStatus = "Success"
-	responseMessage = "[GoogleCloudFunction] :: Message Payload inserted with messageId :: " + parsedJSON.Get("id").String()
-	return http.StatusOK, respStatus, responseMessage
-}
-
-func (producer *GoogleCloudFunctionProducer) invokeGen2Functions(destConfig *Config, parsedJSON gjson.Result) (statusCode int, respStatus, responseMessage string) {
-	ctx := context.Background()
-
-	jsonBytes := []byte(parsedJSON.Raw)
-
 	// Create a POST request
-	req, err := http.NewRequest(http.MethodPost, destConfig.FunctionUrl, strings.NewReader(string(jsonBytes)))
+	req, err := http.NewRequest(http.MethodPost, producer.config.FunctionUrl, bytes.NewReader(jsonData))
 	if err != nil {
-		pkgLogger.Errorf("Failed to create httpRequest for Gen2 Fn: %w", err)
-		return http.StatusBadRequest, "Failure", fmt.Sprintf("[GoogleCloudFunction] Failed to create httpRequest for Gen2 Fn: %s", err.Error())
+		pkgLogger.Errorf("Failed to create httpRequest for Fn: %w", err)
+		return http.StatusBadRequest, "Failure", fmt.Sprintf("[GoogleCloudFunction] Failed to create httpRequest for Fn: %s", err.Error())
 	}
 
 	// Set the appropriate headers
 	req.Header.Set("Content-Type", "application/json")
-	if destConfig.shouldGenerateToken() {
-		token, err := producer.client.GetToken(ctx, destConfig.FunctionUrl, option.WithCredentialsJSON([]byte(destConfig.Credentials)))
+	if producer.config.shouldGenerateToken() {
+		err := producer.config.generateToken(context.Background(), producer.client)
 		if err != nil {
 			pkgLogger.Errorf("failed to receive token: %w", err)
 			return http.StatusUnauthorized, "Failure", fmt.Sprintf("[GoogleCloudFunction] Failed to receive token: %s", err.Error())
 		}
-		destConfig.Token = token
-		destConfig.TokenCreatedAt = time.Now()
-		req.Header.Add("Authorization", "Bearer "+token.AccessToken)
+	}
+	if producer.config.RequireAuthentication && producer.config.Token != nil {
+		req.Header.Set("Authorization", "Bearer "+producer.config.Token.AccessToken)
 	}
 
 	// Make the request using the client
-	resp, err := http.DefaultClient.Do(req)
+	resp, err := producer.httpClient.Do(req)
+	if err != nil && os.IsTimeout(err) {
+		return http.StatusAccepted, "Success", "[GoogleCloudFunction] :: Function is called"
+	}
+
 	var responseBody []byte
 	defer resp.Body.Close()
 	if err == nil {
 		responseBody, err = io.ReadAll(resp.Body)
 	}
 	if err != nil {
 		responseMessage = err.Error()
-		statusCode = http.StatusBadRequest
-		if errors.Is(err, context.DeadlineExceeded) {
-			statusCode = http.StatusGatewayTimeout
-		}
 		respStatus = "Failure"
 		responseMessage = "[GOOGLE_CLOUD_FUNCTION] error :: Function call was not executed " + responseMessage
-		pkgLogger.Errorf("error while calling the Gen2 function :: %v", err)
-		return statusCode, respStatus, responseMessage
+		pkgLogger.Errorf("error while calling the function :: %v", err)
+		return http.StatusBadRequest, respStatus, responseMessage
 	}
 
 	if resp.StatusCode == http.StatusOK {
```

Comment:
Is the check needed for `resp != nil` ?

Reply:
need at first reference too ?
https://github.com/rudderlabs/rudder-server/pull/3903/files#diff-c3fb97248613bb7bbd8ac75a3bd8ae97a314b673e78cb62d406cd2d8816bf1f3R147

Reply:
No, that is not needed.

Reply:
when err is nil then response will not be nil